File: src/services/CockpitService.cpp
#include "CockpitService.h"

#include "domain/AssessmentStatus.h"
#include "domain/Measure.h"
#include "domain/MeasureStatus.h"
#include "domain/Requirement.h"
#include "domain/Standard.h"
#include "domain/TargetObject.h"
#include "services/ReportService.h"

#include <QCoreApplication>
#include <QDate>
#include <QHash>
#include <QStringList>
#include <algorithm>

CockpitService::CockpitService(ICatalogRepository &catalog,
                               IProjectRepository &project,
                               ITargetObjectRepository &targetObjects,
                               IMeasureRepository &measures)
    : m_catalog(catalog)
    , m_project(project)
    , m_targetObjects(targetObjects)
    , m_measures(measures)
{
}

bool CockpitService::responsibleMatches(const QString &responsible, const QString &name,
                                        const QString &email, const QString &needle)
{
    const QString responsibleLower = responsible.trimmed().toLower();
    if (responsibleLower.isEmpty())
        return false;
    if (!needle.trimmed().isEmpty())
        return responsibleLower.contains(needle.trimmed().toLower());
    if (!name.trimmed().isEmpty() && responsibleLower.contains(name.trimmed().toLower()))
        return true;
    if (!email.trimmed().isEmpty() && responsibleLower.contains(email.trimmed().toLower()))
        return true;
    return false;
}

QList<CockpitItem> CockpitService::buildItems(int projectId, const QString &catalogVersion) const
{
    QList<CockpitItem> items;
    ReportService report(m_catalog, m_project, m_targetObjects, m_measures);
    const QList<ReportRow> rows = report.buildSollIstReport(projectId, 0, catalogVersion);

    QHash<QString, ReportRow> assessmentByKey;
    for (const ReportRow &row : rows) {
        CockpitItem item;
        item.kind = CockpitKind::Assessment;
        item.targetObjectId = row.targetObjectId;
        item.targetObjectName = row.targetObjectName;
        item.bausteinDbId = row.bausteinDbId;
        item.bausteinExternalId = row.bausteinExternalId;
        item.requirementDbId = row.requirementDbId;
        item.requirementExternalId = row.requirementExternalId;
        item.title = row.requirementTitle;
        item.statusText = assessmentStatusToString(row.status);
        item.responsible = row.responsible;
        item.dueDate = row.dueDate;
        item.overdue = row.overdue;
        item.assessmentStatus = row.status;
        items.append(item);
        assessmentByKey.insert(QStringLiteral("%1:%2").arg(row.targetObjectId).arg(row.requirementDbId),
                               row);
    }

    QHash<int, Requirement> requirementById;
    const QList<Requirement> requirements =
        m_catalog.loadAllRequirements(StandardType::ITGrundschutz, catalogVersion);
    for (const Requirement &requirement : requirements)
        requirementById.insert(requirement.id, requirement);

    QHash<int, QString> targetNameById;
    const QList<TargetObject> targets = m_targetObjects.loadTargetObjects(projectId);
    for (const TargetObject &target : targets)
        targetNameById.insert(target.id, target.name);

    const QList<Measure> measures = m_measures.loadProjectMeasures(projectId);
    for (const Measure &measure : measures) {
        CockpitItem item;
        item.kind = CockpitKind::Measure;
        item.measureId = measure.id;
        item.targetObjectId = measure.targetObjectId;
        item.targetObjectName = targetNameById.value(measure.targetObjectId,
                                                     QString::number(measure.targetObjectId));
        item.requirementDbId = measure.requirementDbId;
        const Requirement requirement = requirementById.value(measure.requirementDbId);
        if (requirement.id != 0) {
            item.bausteinDbId = requirement.bausteinDbId;
            item.bausteinExternalId = requirement.bausteinExternalId;
            item.requirementExternalId = requirement.externalId;
        }
        const QString key =
            QStringLiteral("%1:%2").arg(measure.targetObjectId).arg(measure.requirementDbId);
        if (assessmentByKey.contains(key)) {
            const ReportRow &linked = assessmentByKey.value(key);
            item.assessmentStatus = linked.status;
            if (item.bausteinDbId == 0) {
                item.bausteinDbId = linked.bausteinDbId;
                item.bausteinExternalId = linked.bausteinExternalId;
                item.requirementExternalId = linked.requirementExternalId;
            }
        }
        item.title = measure.title;
        item.statusText = measureStatusToString(measure.status);
        item.responsible = measure.responsible;
        item.dueDate = measure.dueDate;
        item.measureStatus = measure.status;
        item.overdue = measure.dueDate.isValid() && measure.dueDate < QDate::currentDate()
                       && measure.status != MeasureStatus::Done;
        items.append(item);
    }

    std::sort(items.begin(), items.end(), [](const CockpitItem &left, const CockpitItem &right) {
        if (left.overdue != right.overdue)
            return left.overdue;
        const QDate leftDue = left.dueDate.isValid() ? left.dueDate : QDate(9999, 12, 31);
        const QDate rightDue = right.dueDate.isValid() ? right.dueDate : QDate(9999, 12, 31);
        if (leftDue != rightDue)
            return leftDue < rightDue;
        if (left.targetObjectName != right.targetObjectName)
            return left.targetObjectName < right.targetObjectName;
        if (left.bausteinExternalId != right.bausteinExternalId)
            return left.bausteinExternalId < right.bausteinExternalId;
        if (static_cast<int>(left.kind) != static_cast<int>(right.kind))
            return static_cast<int>(left.kind) < static_cast<int>(right.kind);
        if (left.requirementExternalId != right.requirementExternalId)
            return left.requirementExternalId < right.requirementExternalId;
        return left.title < right.title;
    });

    return items;
}

QList<CockpitItem> CockpitService::applyFilter(const QList<CockpitItem> &items,
                                               const CockpitFilter &filter)
{
    QList<CockpitItem> result;
    for (const CockpitItem &item : items) {
        if (filter.kind == CockpitKindFilter::Assessments && item.kind != CockpitKind::Assessment)
            continue;
        if (filter.kind == CockpitKindFilter::Measures && item.kind != CockpitKind::Measure)
            continue;
        if (filter.hideDone && cockpitItemIsDone(item))
            continue;
        switch (filter.due) {
        case CockpitDueFilter::Overdue:
            if (!item.overdue)
                continue;
            break;
        case CockpitDueFilter::ThisWeek:
            if (!isDueThisWeek(item.dueDate))
                continue;
            break;
        case CockpitDueFilter::HasDate:
            if (!item.dueDate.isValid())
                continue;
            break;
        case CockpitDueFilter::NoDate:
            if (item.dueDate.isValid())
                continue;
            break;
        case CockpitDueFilter::All:
            break;
        }
        if (filter.mineOnly) {
            if (!responsibleMatches(item.responsible, filter.currentUserName, filter.currentUserEmail,
                                    {}))
                continue;
        } else if (!filter.responsibleNeedle.trimmed().isEmpty()) {
            if (!responsibleMatches(item.responsible, {}, {}, filter.responsibleNeedle))
                continue;
        }
        result.append(item);
    }
    return result;
}

CockpitSummary CockpitService::summarize(const QList<CockpitItem> &items)
{
    CockpitSummary summary;
    summary.totalCount = items.size();
    for (const CockpitItem &item : items) {
        if (item.kind == CockpitKind::Assessment)
            ++summary.assessmentCount;
        else
            ++summary.measureCount;
        if (item.overdue)
            ++summary.overdueCount;
        if (isDueThisWeek(item.dueDate))
            ++summary.dueThisWeekCount;
    }
    return summary;
}

QString CockpitService::formatSummary(const CockpitSummary &summary)
{
    if (summary.totalCount == 0)
        return QCoreApplication::translate("CockpitService", "Keine offenen Aufgaben");

    QString text = QCoreApplication::translate("CockpitService", "%1 Einträge")
                       .arg(summary.totalCount);
    QStringList parts;
    if (summary.assessmentCount > 0)
        parts << QCoreApplication::translate("CockpitService", "%1 Bewertungen")
                     .arg(summary.assessmentCount);
    if (summary.measureCount > 0)
        parts << QCoreApplication::translate("CockpitService", "%1 Maßnahmen")
                     .arg(summary.measureCount);
    if (summary.overdueCount > 0)
        parts << QCoreApplication::translate("CockpitService", "%1 überfällig")
                     .arg(summary.overdueCount);
    if (summary.dueThisWeekCount > 0)
        parts << QCoreApplication::translate("CockpitService", "%1 diese Woche")
                     .arg(summary.dueThisWeekCount);
    if (!parts.isEmpty())
        text += QStringLiteral(" – ") + parts.join(QStringLiteral(", "));
    return text;
}

File: Delphi/Services/CockpitService.pas
unit CockpitService;

interface

uses
  System.SysUtils, System.DateUtils, System.Generics.Collections, System.Generics.Defaults,
  IsmsDomain, RepositoryBase, ReportService;

type
  TCockpitService = class
  private
    FCatalog: TCatalogRepositoryBase;
    FProject: TProjectRepositoryBase;
    FTarget: TTargetObjectRepositoryBase;
    FMeasure: TMeasureRepositoryBase;
    class function ItemDueSortKey(const AItem: TCockpitItem): TDateTime;
    class function CompareItems(const L, R: TCockpitItem): Integer;
    class function ResponsibleMatches(const AResponsible, AName, AEmail,
      ANeedle: string): Boolean;
  public
    constructor Create(ACatalog: TCatalogRepositoryBase; AProject: TProjectRepositoryBase;
      ATarget: TTargetObjectRepositoryBase; AMeasure: TMeasureRepositoryBase);
    function BuildItems(AProjectId: Integer;
      const ACatalogVersion: string): TArray<TCockpitItem>;
    class function ApplyFilter(const AItems: TArray<TCockpitItem>;
      const AFilter: TCockpitFilter): TArray<TCockpitItem>;
    class function Summarize(const AItems: TArray<TCockpitItem>): TCockpitSummary;
    class function FormatSummary(const ASummary: TCockpitSummary): string;
  end;

implementation

constructor TCockpitService.Create(ACatalog: TCatalogRepositoryBase;
  AProject: TProjectRepositoryBase; ATarget: TTargetObjectRepositoryBase;
  AMeasure: TMeasureRepositoryBase);
begin
  inherited Create;
  FCatalog := ACatalog;
  FProject := AProject;
  FTarget := ATarget;
  FMeasure := AMeasure;
end;

class function TCockpitService.ItemDueSortKey(const AItem: TCockpitItem): TDateTime;
begin
  if IsValidDate(AItem.DueDate) then
    Exit(AItem.DueDate);
  Result := EncodeDate(9999, 12, 31);
end;

class function TCockpitService.CompareItems(const L, R: TCockpitItem): Integer;
begin
  if L.Overdue <> R.Overdue then
  begin
    if L.Overdue then
      Exit(-1);
    Exit(1);
  end;
  Result := CompareDate(ItemDueSortKey(L), ItemDueSortKey(R));
  if Result <> 0 then
    Exit;
  Result := CompareText(L.TargetObjectName, R.TargetObjectName);
  if Result <> 0 then
    Exit;
  Result := CompareText(L.BausteinExternalId, R.BausteinExternalId);
  if Result <> 0 then
    Exit;
  Result := Ord(L.Kind) - Ord(R.Kind);
  if Result <> 0 then
    Exit;
  Result := CompareText(L.RequirementExternalId, R.RequirementExternalId);
  if Result <> 0 then
    Exit;
  Result := CompareText(L.Title, R.Title);
end;

class function TCockpitService.ResponsibleMatches(const AResponsible, AName, AEmail,
  ANeedle: string): Boolean;
var
  ResponsibleLower, NeedleLower: string;
begin
  ResponsibleLower := LowerCase(Trim(AResponsible));
  if ResponsibleLower = '' then
    Exit(False);
  if Trim(ANeedle) <> '' then
  begin
    NeedleLower := LowerCase(Trim(ANeedle));
    Exit(Pos(NeedleLower, ResponsibleLower) > 0);
  end;
  if (Trim(AName) <> '') and (Pos(LowerCase(Trim(AName)), ResponsibleLower) > 0) then
    Exit(True);
  if (Trim(AEmail) <> '') and (Pos(LowerCase(Trim(AEmail)), ResponsibleLower) > 0) then
    Exit(True);
  Result := False;
end;

function TCockpitService.BuildItems(AProjectId: Integer;
  const ACatalogVersion: string): TArray<TCockpitItem>;
var
  Report: TReportService;
  Rows: TArray<TReportRow>;
  Row: TReportRow;
  Measures: TArray<TMeasure>;
  Measure: TMeasure;
  Requirements: TArray<TRequirement>;
  Requirement: TRequirement;
  Targets: TArray<TTargetObject>;
  Target: TTargetObject;
  ReqById: TDictionary<Integer, TRequirement>;
  TargetNameById: TDictionary<Integer, string>;
  AssessmentByKey: TDictionary<string, TReportRow>;
  Item: TCockpitItem;
  List: TList<TCockpitItem>;
  Key: string;
  Linked: TReportRow;
begin
  List := TList<TCockpitItem>.Create;
  ReqById := TDictionary<Integer, TRequirement>.Create;
  TargetNameById := TDictionary<Integer, string>.Create;
  AssessmentByKey := TDictionary<string, TReportRow>.Create;
  Report := TReportService.Create(FCatalog, FProject, FTarget, FMeasure);
  try
    Rows := Report.BuildSollIstReport(AProjectId, 0, ACatalogVersion);
    for Row in Rows do
    begin
      FillChar(Item, SizeOf(Item), 0);
      Item.Kind := ckAssessment;
      Item.TargetObjectId := Row.TargetObjectId;
      Item.TargetObjectName := Row.TargetObjectName;
      Item.BausteinDbId := Row.BausteinDbId;
      Item.BausteinExternalId := Row.BausteinExternalId;
      Item.RequirementDbId := Row.RequirementDbId;
      Item.RequirementExternalId := Row.RequirementExternalId;
      Item.Title := Row.RequirementTitle;
      Item.StatusText := AssessmentStatusToString(Row.Status);
      Item.Responsible := Row.Responsible;
      Item.DueDate := Row.DueDate;
      Item.Overdue := Row.Overdue;
      Item.AssessmentStatus := Row.Status;
      List.Add(Item);
      AssessmentByKey.AddOrSetValue(Format('%d:%d', [Row.TargetObjectId, Row.RequirementDbId]), Row);
    end;

    Requirements := FCatalog.LoadAllRequirements(stITGrundschutz, ACatalogVersion);
    for Requirement in Requirements do
      ReqById.AddOrSetValue(Requirement.Id, Requirement);

    Targets := FTarget.LoadTargetObjects(AProjectId);
    for Target in Targets do
      TargetNameById.AddOrSetValue(Target.Id, Target.Name);

    Measures := FMeasure.LoadProjectMeasures(AProjectId);
    for Measure in Measures do
    begin
      FillChar(Item, SizeOf(Item), 0);
      Item.Kind := ckMeasure;
      Item.MeasureId := Measure.Id;
      Item.TargetObjectId := Measure.TargetObjectId;
      if not TargetNameById.TryGetValue(Measure.TargetObjectId, Item.TargetObjectName) then
        Item.TargetObjectName := IntToStr(Measure.TargetObjectId);
      Item.RequirementDbId := Measure.RequirementDbId;
      if ReqById.TryGetValue(Measure.RequirementDbId, Requirement) then
      begin
        Item.BausteinDbId := Requirement.BausteinDbId;
        Item.BausteinExternalId := Requirement.BausteinExternalId;
        Item.RequirementExternalId := Requirement.ExternalId;
      end;
      Key := Format('%d:%d', [Measure.TargetObjectId, Measure.RequirementDbId]);
      if AssessmentByKey.TryGetValue(Key, Linked) then
      begin
        Item.AssessmentStatus := Linked.Status;
        if Item.BausteinDbId = 0 then
        begin
          Item.BausteinDbId := Linked.BausteinDbId;
          Item.BausteinExternalId := Linked.BausteinExternalId;
          Item.RequirementExternalId := Linked.RequirementExternalId;
        end;
      end;
      Item.Title := Measure.Title;
      Item.StatusText := MeasureStatusToString(Measure.Status);
      Item.Responsible := Measure.Responsible;
      Item.DueDate := Measure.DueDate;
      Item.MeasureStatus := Measure.Status;
      Item.Overdue := IsValidDate(Measure.DueDate) and (Measure.DueDate < Date) and
        (Measure.Status <> msDone);
      List.Add(Item);
    end;

    List.Sort(TComparer<TCockpitItem>.Construct(
      function(const L, R: TCockpitItem): Integer
      begin
        Result := CompareItems(L, R);
      end));
    Result := List.ToArray;
  finally
    Report.Free;
    AssessmentByKey.Free;
    TargetNameById.Free;
    ReqById.Free;
    List.Free;
  end;
end;

class function TCockpitService.ApplyFilter(const AItems: TArray<TCockpitItem>;
  const AFilter: TCockpitFilter): TArray<TCockpitItem>;
var
  Item: TCockpitItem;
  List: TList<TCockpitItem>;
begin
  List := TList<TCockpitItem>.Create;
  try
    for Item in AItems do
    begin
      case AFilter.Kind of
        ckfAssessments:
          if Item.Kind <> ckAssessment then
            Continue;
        ckfMeasures:
          if Item.Kind <> ckMeasure then
            Continue;
      end;
      if AFilter.HideDone and CockpitItemIsDone(Item) then
        Continue;
      case AFilter.Due of
        cdfOverdue:
          if not Item.Overdue then
            Continue;
        cdfThisWeek:
          if not IsDueThisWeek(Item.DueDate) then
            Continue;
        cdfHasDate:
          if not IsValidDate(Item.DueDate) then
            Continue;
        cdfNoDate:
          if IsValidDate(Item.DueDate) then
            Continue;
      end;
      if AFilter.MineOnly then
      begin
        if not ResponsibleMatches(Item.Responsible, AFilter.CurrentUserName,
          AFilter.CurrentUserEmail, '') then
          Continue;
      end
      else if Trim(AFilter.ResponsibleNeedle) <> '' then
      begin
        if not ResponsibleMatches(Item.Responsible, '', '', AFilter.ResponsibleNeedle) then
          Continue;
      end;
      List.Add(Item);
    end;
    Result := List.ToArray;
  finally
    List.Free;
  end;
end;

class function TCockpitService.Summarize(const AItems: TArray<TCockpitItem>): TCockpitSummary;
var
  Item: TCockpitItem;
begin
  FillChar(Result, SizeOf(Result), 0);
  Result.TotalCount := Length(AItems);
  for Item in AItems do
  begin
    case Item.Kind of
      ckAssessment: Inc(Result.AssessmentCount);
      ckMeasure: Inc(Result.MeasureCount);
    end;
    if Item.Overdue then
      Inc(Result.OverdueCount);
    if IsDueThisWeek(Item.DueDate) then
      Inc(Result.DueThisWeekCount);
  end;
end;

class function TCockpitService.FormatSummary(const ASummary: TCockpitSummary): string;
var
  Parts: TArray<string>;
begin
  if ASummary.TotalCount = 0 then
    Exit('Keine offenen Aufgaben');
  Result := Format('%d Eintr'#$00E4'ge', [ASummary.TotalCount]);
  SetLength(Parts, 0);
  if ASummary.AssessmentCount > 0 then
  begin
    SetLength(Parts, Length(Parts) + 1);
    Parts[High(Parts)] := Format('%d Bewertungen', [ASummary.AssessmentCount]);
  end;
  if ASummary.MeasureCount > 0 then
  begin
    SetLength(Parts, Length(Parts) + 1);
    Parts[High(Parts)] := Format('%d Ma'#$00DF'nahmen', [ASummary.MeasureCount]);
  end;
  if ASummary.OverdueCount > 0 then
  begin
    SetLength(Parts, Length(Parts) + 1);
    Parts[High(Parts)] := Format('%d '#252'berf'#$00E4'llig', [ASummary.OverdueCount]);
  end;
  if ASummary.DueThisWeekCount > 0 then
  begin
    SetLength(Parts, Length(Parts) + 1);
    Parts[High(Parts)] := Format('%d diese Woche', [ASummary.DueThisWeekCount]);
  end;
  if Length(Parts) > 0 then
    Result := Result + ' ' + #$2013 + ' ' + string.Join(', ', Parts);
end;

end.
